packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -1125,16 +1152,48 @@ void Add(
           }
         }
 
-        if (C->isFillComplete ()) {
-          C->sumIntoGlobalValues (globalRow, numEntries, 
+        if (CwasFillComplete) {
+          size_t result = C->sumIntoGlobalValues (globalRow, numEntries, 
                                   reinterpret_cast<Scalar *>(Values.data()), Indices.data());
+          TEUCHOS_TEST_FOR_EXCEPTION(result != numEntries, std::logic_error,
+              prefix << "sumIntoGlobalValues failed to add entries from A or B into C.");
         } else {
           C->insertGlobalValues (globalRow,  numEntries, 
                                  reinterpret_cast<Scalar *>(Values.data()), Indices.data());
         }
       }
     }
   }
+  if(CwasFillComplete) {
+    // This version of fillComplete will reuse the domain
+    // and range maps from the previous fillComplete.
+    C->fillComplete();
+  }
+}
+
+// This version of Add takes C as const RCP&, so C must not be null on input (in this case,
+// it is allocated and constructed in this function). Otherwise, its behavior is identical
```

Comment:
This phrasing is a little weird to me.  Because C is a `const RCP&`, it can't be allocated in this version.

Reply:
Good catch, that was copied from the description of the ``RCP&`` version.